src/rgw/rgw_string.cc

```diff
@@ -2,44 +2,20 @@
 // vim: ts=8 sw=2 smarttab ft=cpp
 
 #include "rgw_string.h"
-
-static bool char_eq(char c1, char c2)
-{
-  return c1 == c2;
-}
-
-static bool ci_char_eq(char c1, char c2)
-{
-  return tolower(c1) == tolower(c2);
-}
+#include <fnmatch.h>
 
 bool match_wildcards(std::string_view pattern, std::string_view input,
                      uint32_t flags)
 {
-  const auto eq = (flags & MATCH_CASE_INSENSITIVE) ? &ci_char_eq : &char_eq;
+  bool case_insensive = flags & MATCH_CASE_INSENSITIVE;
+  uint32_t  flag = 0;
 
-  auto it1 = pattern.begin();
-  auto it2 = input.begin();
-  while (true) {
-    if (it1 == pattern.end())
-      return it2 == input.end();
-    if (*it1 == '*') {
-      if (it1 + 1 == pattern.end())
-        return true;
-      if (it2 == input.end() || eq(*(it1 + 1), *it2))
-        ++it1;
-      else
-        ++it2;
-      continue;
-    }
-    if (it2 == input.end())
-      return false;
-    if (*it1 == '?' || eq(*it1, *it2)) {
-      ++it1;
-      ++it2;
-      continue;
-    }
+  if (case_insensive) {
+    flag = FNM_CASEFOLD;
+  }
+  if (fnmatch(pattern.data(), input.data(), flag) == 0) {
```

Comment:
unfortunately, we can't assume that these string_views are null-terminated

Reply:
here  we change  string_views to string  maybe OK ？

Reply:
> unfortunately, we can't assume that these string_views are null-terminated

in http://web.mit.edu/freebsd/csup/fnmatch.c see

#define	EOS	'\0'

static const char *rangematch(const char *, char, int);

int
fnmatch(const char *pattern, const char *string, int flags)
{
	const char *stringstart;
	char c, test;

	for (stringstart = string;;)
		switch (c = *pattern++) {
		case EOS:
			if ((flags & FNM_LEADING_DIR) && *string == '/')
				return (0);
			return (*string == EOS ? 0 : FNM_NOMATCH);
		case '?':

Reply:
well for me 1) maybe; 2) probably not?

Reply:
is fnmatch() is a good fit for this, or should we extend the existing matching logic? @mattbenjamin @adamemerson

Reply:
change it to string ?

Reply:
change it to const std::string in match_wildcards()?

Reply:
This should be fine. I wanted to check that there wasn't a good way to make Boost Regex greedier, but there doesn't seem to be.

Reply:
Though I think I'd prefer extending the existing logic to extra allocations.

Reply:
> Though I think I'd prefer extending the existing logic to extra allocations.

@adamemerson @mattbenjamin  who will extend the existing logic to extra allocations ?

Reply:
@cbodley  @adamemerson  @mattbenjamin  
What should we do next ?
Maybe one of us  should extend the existing logic to extra allocations.

Reply:
@qiuxinyidian sorry for the delays. i do think we should match aws' behavior for this parsing, and i'd rather not make this parsing too complicated to in the process. so i do like your approach to use fnmatch()

> if we are going to switch to `fnmatch()`, then the arguments to `match_wildcards()` and `match_policy()` should change to `const char*`s or `const std::string&`s to reflect that requirement

can you try this way?

Reply:
@cbodley OK, i will try

Reply:
@cbodley  src/rgw/rgw_iam_policy.cc has little confilict ,can you help me fix it?